diffusion/models/transformer.py

```diff
@@ -0,0 +1,444 @@
+        self.k_norm = nn.LayerNorm(self.num_features, elementwise_affine=True, eps=1e-6)
+        # Initialize all biases to zero
+        nn.init.zeros_(self.qkv.bias)
+        # Init the standard deviation of the weights to 0.02 as is tradition
+        nn.init.normal_(self.qkv.weight, std=0.02)
+
+    def forward(self, x: torch.Tensor, t: torch.Tensor) -> Tuple[torch.Tensor, torch.Tensor, torch.Tensor]:
+        # Calculate the modulations
+        mods = self.adaLN_mlp(t).unsqueeze(1).chunk(2, dim=2)
+        # Forward, with modulations
+        x = modulate(self.input_norm(x), mods[0], mods[1])
+        # Calculate the query, key, and values all in one go
+        q, k, v = self.qkv(x).chunk(3, dim=-1)
+        q = self.q_norm(q)
+        k = self.k_norm(k)
+        return q, k, v
+
+
+class SelfAttention(nn.Module):
+    """Standard multihead self attention layer that supports masking.
+
+    Args:
+        num_features (int): Number of input features.
+        num_heads (int): Number of attention heads.
+    """
+
+    def __init__(self, num_features: int, num_heads: int):
+        super().__init__()
+        self.num_features = num_features
+        self.num_heads = num_heads
+
+    def forward(self,
+                q: torch.Tensor,
+                k: torch.Tensor,
+                v: torch.Tensor,
+                mask: Optional[torch.Tensor] = None) -> torch.Tensor:
+        # Get the shape of the inputs
+        B, T, C = v.size()
+        # Reshape the query, key, and values for multi-head attention
+        q = q.view(B, T, self.num_heads, C // self.num_heads).transpose(1, 2)
+        k = k.view(B, T, self.num_heads, C // self.num_heads).transpose(1, 2)
+        v = v.view(B, T, self.num_heads, C // self.num_heads).transpose(1, 2)
+        # Native torch attention
+        attention_out = F.scaled_dot_product_attention(q, k, v, attn_mask=mask)  # (B, H, T, C/H)
+        # Swap the sequence length and the head dimension back and get rid of num_heads.
+        attention_out = attention_out.transpose(1, 2).contiguous().view(B, T, C)  # (B, T, C)
+        return attention_out
+
+
+class PostAttentionBlock(nn.Module):
+    """Block to postprocess v after attention.
+
+    Includes adaptive layernorms.
+
+    Args:
+        num_features (int): Number of input features.
+        expansion_factor (int): Expansion factor for the MLP. Default: `4`.
+    """
+
+    def __init__(self, num_features: int, expansion_factor: int = 4):
+        super().__init__()
+        self.num_features = num_features
+        self.expansion_factor = expansion_factor
+        # AdaLN MLP for post-attention. Initialized to zero so modulation acts as identity at initialization.
+        self.adaLN_mlp_linear = nn.Linear(self.num_features, 4 * self.num_features, bias=True)
+        nn.init.zeros_(self.adaLN_mlp_linear.weight)
+        nn.init.zeros_(self.adaLN_mlp_linear.bias)
+        self.adaLN_mlp = nn.Sequential(nn.SiLU(), self.adaLN_mlp_linear)
```

Comment:
probably refactor into something like 

```
AdaptiveLayerNorm(num_features, expansion_factor)
```
same for the one in PreAttentionBlock

Reply:
Added an `AdaptiveLayerNorm` and also a `ModulationLayer` to replace the duplicated AdaLN logic in the Pre/Post attention blocks.